### ISSUE TITLE
7.3.3 (4/5) naming the current instance

### ISSUE THREAD

**Issue by nholsti** [pat-rogers/Ada-202x-WG9-Informal-Review#149]
The phrase "a name that denotes the declaration" seems suspect, because RM 8.3 (16) says that a declaration is hidden from visibility until the end of the declaration, except in some cases which do not seem to include the declaration of a private type or a private extension. If the declaration of a private type or private extension is hidden until the end of the declaration, it is hidden within the Default_Initial_Condition expression, and therefore that expression cannot contain a name that denotes the declaration of the private type or private extension. Or so it seems to me.

Note that the analogous definition of Type_Invariant in RM 7.3.2 (5/4) uses a different phrase: "the identifier of the first subtype of the associated type". Why is the wording different in these two places, and which is correct?

**Comment by sttaft** [comment]
This should be fixed in RM 8.3(16) to indicate that a declaration is visible within its aspect specification, if any.

**Comment by ARG-Editor** [comment]
I agree with Tucker. I've left this marked as a bug for now, since this seems to be a conflict with any aspect (such as a predicate, type invariant, etc.) that uses a name from a declaration in the aspect specification. We need a wording proposal.

**Comment by sttaft** [comment]
Proposed change to RM 8.3(16):

* A declaration is hidden from all visibility until the end of the declaration, except: 
{   * A declaration is visible within the aspect_specification of the declaration, if any;
}

**Comment by ARG-Editor** [comment]
Humm. The other bullets here talk about the limits to a declaration from being hidden from all visibility. I don't think we want to talk here about something being visible, since that might be construed as overriding other visibility rules.

More consistent would be:
    {* The declaration is only hidden from all visibility until the reserved word **with** of the aspect_specification of the declaration, if any.}

But that's a little weird, as there's no conditionally to it. Other ideas?

**Comment by ARG-Editor** [comment]
Maybe it would be better to just say this as part of paragraph 16:

A declaration is hidden from all visibility until {the reserved word with of the aspect_specification of the declaration, if any, or }the end of the declaration, except: 

This does assume that aspect_specifications of declarations are last (or at least after anything where visibility matters; "is null" or "is abstract" is not interesting that way). Any counter cases??

**Comment by sttaft** [comment]
Steve Baird writes:
If we are going to take this approach, I suggest adding two more words,
"otherwise until", as in:

   A declaration is hidden from all visibility until {either the reserved
   word with of the aspect_specification of the declaration, if any, or
   otherwise until} the end of the declaration, except:

But do we need to say anything at all here? It seems like we don't care
whether the declaration is hidden from all visibility within its
aspect specifications because of 13.1.1(11/5), so no action of any kind
is needed.

**Comment by sttaft** [comment]
I don't understand the desire to use a lexeme for this rule, when we already have a nice syntactic category defined.  I can see in other cases why we resort to specifying things by particular lexemes like "is" but that is because there might not be a convenient syntactic category to use, but here we have a zillion different cases where an aspect_specification can appear, with some of the nearby syntax rules using "with" for other things (e.g. type extensions), so I would argue pretty strongly for using the syntactic category rather than the lexeme "with."  And as you point out, it presumes they come last.

Steve's question about 13.1.1(11/5) is interesting, saying that all name resolution is delayed, but that doesn't really make any sense when talking about the current instance, parameters, etc., since none of those are directly visible when we are at the end of the enclosing declaration list.  To be completely honest, we should say that those sorts of names (i.e. names of parts of the associated declaration), it makes no sense to delay name resolution.  But that is opening a different can of worms!

In any case, to address Randy's original concern, I would propose the following (I am providing more context to understand Randy's original concern about consistency):

Modify the set of bullets starting with RM 8.3(16):

> * A declaration is hidden from all visibility until the end of the declaration, except: {
> 
>   * A declaration is not hidden from all visibility within its aspect_specification, if any;}
>   * For a record type or record extension, the declaration is hidden from all visibility only until the reserved word record;
> 
>   * For a package_declaration, generic_package_declaration, subprogram_body, or expression_function_declaration, the declaration is hidden from all visibility only until the reserved word is of the declaration; 
> 
>   * For a task declaration or protected declaration, the declaration is hidden from all visibility only until the reserved word with of the declaration if there is one, or the reserved word is of the declaration if there is no with.

**Comment by ARG-Editor** [comment]
My understanding of this "hidden from all visibility" thingy is that it is something that exists initially and then is canceled; for Janus/Ada, that gives the point where the entity is actually declared. (I realize that it is also used in overriding, but there it is just a preference rule, allowing the choice of a particular declaration when you have several equivalent ones). Having "hidden from all visibility" being something that changes multiple times for a declaration is something new and would clearly require some new mechanisms in Janus/Ada (obviously can't speak about other compilers).

Although, as Steve says, it doesn't really matter as no lookup or resolution occurs until the freezing point/end of dec;laration list, and nothing should be hidden from all visibility there (for this reason, other kinds aren't involved in this wording anyway). Tucker's comment about "direct visibility" is irrelevant, that's handled by other rules; the issue would be that "hidden from all visibility" would prevent something from being visible (or directly visible) at all.

The only cases where I'm aware that something would need to be resolved immediately is a few aspects like Pure which override the usual resolution and require immediate resolution. But none of those use local declarations, so I think we're still OK.

So I am leaning toward marking this No Action. 

If Tucker has some other concern that I missed, then I think we should mark this as "Deferred". Changing the visibility rules when we don't have to seems dangerous and should not be done as a "quick fix" unless we're certain there is an actual problem (and we're not).

I'll wait until tomorrow for a definitive answer to let Tucker weight in.

**Comment by ARG-Editor** [comment]
..and then I labeled it anyway. It's easy enough to change if Tucker has some other issue.

**Comment by stephen-baird-751** [comment]
Tuck makes a good point about parameters. The argument I had in mind for how 13.1.1(11/5) applies to this question might work for the current instance, but it breaks down when we talk about things like parameters or components that are declared as part of an enclosing aspect-bearing declaration (e.g., a declaration of a subprogram or of a type), and which need to be nameable in an aspect specification.

**Comment by nholsti** [comment]
Interesting discussions, but no-one has commented on the second part of my original issue, which is that the analogous definition of the current instance for Type_Invariant in RM 7.3.2 (5/4) uses a different phrase: "the identifier of the first subtype of the associated type", instead of "a name that denotes the declaration". The former suggests a direct comparison of identifier strings on the lexical level, as for the identifiers that often follow "end" keywords, while the latter suggests the outcome of name resolution. Is the wording intentionally different? Or would it be better to use the same wording in both places?

I haven't checked all the other similar rules in the RM.

**Comment by ARG-Editor** [comment]
Answering Steve:

The argument doesn't "break down" for components and parameters, it works exactly the same for them. Remember the difference between "visibility" and "direct visibility" when talking formally. To use a name at all in Ada, it has to be "visible". We know that parameters and components are "visible" in the scope of the associated entity, as one can make a named call (parameters) or use selected notation (components) in that scope. (You could place one of those things at the end of the declarative part and it would work.) So it's not necessary to worry about whether those things are "visible".

It's usual to have rules that make things "directly visible" (that are already "visible") in particular scopes (think discriminants and protected components), and we have one for things in an aspect_specification (13.1.1(12/5)). So there is no problem with that.

So this is not an issue. The OPs question was about "hidden-from-all-visibility", which would make 13.1.1(12/5) (and the rule in 8.6 for types as well) ineffective. And the answer that 13.1.1(11/5) says that the resolution is done at the end of the declaration list (and not as part of the declaration) takes care of that. Thus this issue should be "No Action".

**Comment by ARG-Editor** [comment]
Answering Nicolas: This second question is unrelated to the first and thus I've split it into it's own issue, #177.
